### PR TITLE
Send shutdown command to DB

### PR DESCRIPTION
**Pull request by al-rigazzi**
This PR changes the way `Orchestrator`s are stopped.

Currently, the `Orchestrator` process or job is simply killed: this proved not to be a good technique with some WLMs, where the launcher and the launched process are different. In such cases, it is impossible to stop a running process and orchestrators stay up until the allocation is relinquished.

With this PR, we added the execution of an explicit `shutdown` command for each DB shard, which properly terminates the Orchestrator in all cases.

A few tweaks to the current test suite were also implemented as part of this PR, to make sure that executing them through `mpiexec` on some systems did not fail.

### CODE REVIEW

**Review by MattToast — CHANGES_REQUESTED** [review]
Couple of small knit picks, but otherwise LGTM!!

File: tests/backends/test_dbmodel.py [hunk 1]
```diff
@@ -26,17 +26,17 @@
...
 
 import sys
-import time
 
 import pytest
 
-import smartsim
 from smartsim import Experiment, status
 from smartsim._core.utils import installed_redisai_backends
+from smartsim.entity import ensemble
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply tests/backends/test_dbmodel.py:34]: Note to self: never click the first result in the "Quick fix" context menu.

File: smartsim/_core/utils/redis.py [hunk 2]
```diff
@@ -214,3 +214,35 @@ def set_script(db_script: DBScript, client: Client) -> None:
...
+    :raises SmartSimError: if cluster creation fails
+    """
+    for host in hosts:
+        host_ip = get_ip_from_host(host)
+        for port in ports:
+            # call cluster command
+            redis_cli = CONFIG.database_cli
+            cmd = [redis_cli]
+            cmd += ["-h", host_ip, "-p", str(port)]
+            cmd += ["shutdown"]
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/_core/utils/redis.py:239]: Pedantic, but right.

File: smartsim/_core/utils/redis.py [hunk 3]
```diff
@@ -214,3 +214,35 @@ def set_script(db_script: DBScript, client: Client) -> None:
...
+
+    :param hosts: List of hostnames to connect to
+    :type hosts: List[str]
+    :param ports: List of ports for each hostname
+    :type ports: List[int]
+    :raises SmartSimError: if cluster creation fails
+    """
+    for host in hosts:
+        host_ip = get_ip_from_host(host)
+        for port in ports:
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/_core/utils/redis.py:234]: Nice!

File: smartsim/_core/utils/redis.py [hunk 4]
```diff
@@ -214,3 +214,35 @@ def set_script(db_script: DBScript, client: Client) -> None:
...
+            cmd += ["shutdown"]
+            returncode, out, err = execute_cmd(
+                cmd, proc_input="yes", shell=False, timeout=10
+            )
+
+            if returncode != 0:
+                logger.error(out)
+                logger.error(err)
+
+            logger.debug(out)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/_core/utils/redis.py:248]: You mean the `logger.debug(out)` line?

File: smartsim/database/orchestrator.py [hunk 5]
```diff
@@ -407,6 +407,13 @@ def set_hosts(self, host_list: t.List[str]) -> None:
         if self.launcher == "lsf":
             for db in self.dbnodes:
                 db.set_hosts(host_list)
+        elif (self.launcher == "pals"
+              and isinstance(self.dbnodes[0].run_settings, PalsMpiexecSettings)
+              and self.dbnodes[0].is_mpmd
+              and hasattr(self.dbnodes[0].run_settings, "mpmd")):
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/database/orchestrator.py:413]: Correct, indeed

File: smartsim/settings/palsSettings.py [hunk 6]
```diff
@@ -221,3 +221,20 @@ def format_env_vars(self) -> t.List[str]:
...
+        :type host_list: str | list[str]
+        :raises TypeError: if not str or list of str
+        """
+        if isinstance(host_list, str):
+            host_list = [host_list.strip()]
+        if not isinstance(host_list, list):
+            raise TypeError("host_list argument must be a list of strings")
+        if not all(isinstance(host, str) for host in host_list):
+            raise TypeError("host_list argument must be list of strings")
+        self.run_args["hosts"] = ",".join(host_list)
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/settings/palsSettings.py:240]: Yeah, technically `host` works too, but I prefer to comply to their docs.

File: smartsim/settings/palsSettings.py [hunk 7]
```diff
@@ -221,3 +221,20 @@ def format_env_vars(self) -> t.List[str]:
             formatted += ["--envlist", ",".join(export_vars)]
 
         return formatted
+
+    def set_hostlist(self, host_list: t.Union[str, t.List[str]]) -> None:
+        """Set the hostlist for the PALS ``mpirun`` command
```
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/settings/palsSettings.py:226]: Done

File: smartsim/_core/utils/redis.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (al-rigazzi) [reply smartsim/_core/utils/redis.py:247]: Done

**Review by MattToast — APPROVED** [review]
LGTM!! Thanks for adding this; this is definitely a much more clean way of tearing down DBs!!

**Pull request merged**